### PR TITLE
Refactor to facade pattern

### PR DESCRIPTION
**Pull request by nicorithner**
- Create search_facade tests
- Update vcr cassete and search
- Create GifService spec tests
- Create GifService service Added class method search_gifts_api(query) to perform the api call
Parsed raw data's body to json
- Create test for poro Gif
- Create poro Gif with an attribute of image
- Create search facade Facade runs a query and returns an object
- Upgrade search controller to use facade pattern
- Update stories
- Added styling to the container and the gif item
- Change div name from front_page_container to main
- Change spacing
- Started a README and added a gif to demo

### CODE REVIEW

File: app/views/search/index.html.erb [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <div class="results">
       <% @gifs.each do |gif| %>
         <div class="gif">
-        <img src= "<%= gif[:images][:original][:url]%>"> <br/>
+          <img src= "<%= gif.image%>">
```
Comment (hashmaster3k) [inline app/views/search/index.html.erb:7]: Did this end up working with the image_tag too? Just curious, either way is fine.

File: app/controllers/search_controller.rb [hunk 2]
```diff
@@ -1,8 +1,5 @@
 class SearchController < ApplicationController
   def index
-    search_params = params[:q]
-    response = Faraday.get("https://api.giphy.com/v1/gifs/search?api_key=#{ENV['giphy_api_key']}&q=#{search_params}&limit=10")
-    json = JSON.parse(response.body, symbolize_names: true)
-    @gifs = json[:data]
+    @gifs = SearchFacade.search_gifs(params[:q])
   end
```
Comment (hashmaster3k) [inline app/controllers/search_controller.rb:4]: Looks good! Much cleaner.

File: app/facades/search_facade.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class SearchFacade
+  def self.search_gifs(query)
+    results = GifService.search_gifs_api(query) # step 1. runs the query
+    results[:data].map { |data| Gif.new(data) } # step 2. transforms into object
```
Comment (hashmaster3k) [inline app/facades/search_facade.rb:4]: Good use of single line statement. Facade is light and seems like it's doing what it's supposed to.

File: app/services/gif_service.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class GifService
+  def self.search_gifs_api(query)
+    response = Faraday.get("https://api.giphy.com/v1/gifs/search?api_key=#{ENV['giphy_api_key']}&q=#{query}&limit=10")
+    JSON.parse(response.body, symbolize_names: true)
+  end
```
Comment (hashmaster3k) [inline app/services/gif_service.rb:5]: Looks good here too.

File: spec/facades/search_facade_spec.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        gifs = SearchFacade.search_gifs('Duffy Duck')
+        
+        expect(gifs).to be_an(Array)
+        expect(gifs.count).to eq(10)
+        expect(gifs.first).to be_a(Gif)
+        expect(gifs.first.image).to be_a(String)
+      end
+    end
+  end
+end
```
Comment (hashmaster3k) [inline spec/facades/search_facade_spec.rb:15]: Good test! Dynamic and not too specific.

File: spec/poros/gif_spec.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+RSpec.describe Gif do
+  it "exists" do
+      data = {images: {original: {url: 'url string'}}}
```
Comment (hashmaster3k) [inline spec/poros/gif_spec.rb:4]: Glad you got this working.

File: spec/services/gif_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+      VCR.use_cassette('giphy_search') do
+        gifs_list = GifService.search_gifs_api('Duffy Duck')
+
+        expect(gifs_list).to be_an(Hash)
+        expect(gifs_list[:data]).to be_an(Array)
+        expect(gifs_list[:data].first).to be_an(Hash)
+        expect(gifs_list[:data].first[:id]).to be_a(String)
+        expect(gifs_list[:data].first[:type]).to eq('gif')
+        expect(gifs_list[:data].first[:images][:original][:url]).to be_a(String)
+      end
```
Comment (hashmaster3k) [inline spec/services/gif_service_spec.rb:14]: The ID is returned as a string? That's surprising.

**Review by hashmaster3k — APPROVED** [review]
Overall, everything looks great! Seems like are you understanding the facade/service design. Good job!

File: spec/services/gif_service_spec.rb (hunk 7 above)
Comment (nicorithner) [reply spec/services/gif_service_spec.rb:14]: Yeah, that surprised me too but it was a string combining numbers and letters

File: app/views/search/index.html.erb (hunk 1 above)
Comment (nicorithner) [reply app/views/search/index.html.erb:7]: I need to try that. It should!

**Pull request merged**